concordium-smart-contract-testing/src/impls.rs

```diff
@@ -1056,11 +1223,277 @@ impl Chain {
+    ///
+    /// let mut chain = Chain::new();
+    /// chain.set_block_time(Timestamp::from_timestamp_millis(123));
+    /// ```
+    pub fn set_block_time(&mut self, block_time: Timestamp) {
+        self.parameters.block_time = block_time;
+    }
+
+    /// Set the block time by querying the external node.
+    ///
+    /// Use the `block` parameter to specify the block used when querying the
+    /// block time. For consistency between queries, the first block hash
+    /// used will be "sticky", which means that it is saved and used for
+    /// subsequent queries, unless overridden by specifying a `block`. If
+    /// `block == None` for the first query, it will use the *last final
+    /// block* instead. In either case, the block hash will be saved and
+    /// used for future queries.
+    /// To see the "sticky block", use [`sticky_block`][Chain::sticky_block].
+    ///
+    /// The external node must be setup prior to this call via the method
+    /// [`Chain::setup_external_node_connection`], otherwise an error is
+    /// returned.
+    ///
+    /// If the queried block time is prior to January 1, 1970 0:00:00 UTC (aka
+    /// “UNIX timestamp”), a block time of `0` will be used instead.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use concordium_smart_contract_testing::*;
+    /// use std::str::FromStr;
+    ///
+    /// let mut chain = Chain::builder()
+    ///     .external_node_connection("http://node.testnet.concordium.com:20000")
+    ///     .build()
+    ///     .unwrap();
+    ///
+    /// // Use the last final block and set the "sticky block" to it.
+    /// chain.set_block_time_via_external_node(None).unwrap();
+    ///
+    /// // Use a specific block instead of the sticky block.
+    /// // This does not change the sticky block.
+    /// chain
+    ///     .set_block_time_via_external_node(Some(
+    ///         BlockHash::from_str("95ff82f26892a2327c3e7ac582224a54d75c367341fbff209bce552d81349eb0")
+    ///             .unwrap(),
+    ///     ))
+    ///     .unwrap();
+    /// ```
+    pub fn set_block_time_via_external_node(
+        &mut self,
+        block: Option<BlockHash>,
+    ) -> Result<(), ExternalNodeError> {
+        let connection = self.external_node_connection()?;
+
+        // Get the timestamp in milliseconds *before or after* the unix epoch from the
+        // external node.
+        let timestamp_before_or_after =
+            connection.with_client(block, |block_identifier, mut client| async move {
+                let response = client.get_block_info(block_identifier).await?;
+                let block_time = response.response.block_slot_time.timestamp_millis();
+                Ok((response.block_hash, block_time))
+            })?;
+
+        // Our `Timestamp` only supports times *after* the unix epoch, so we use `0` if
+        // the queried timestamp is negative.
+        let timestamp_after: u64 = i64::max(timestamp_before_or_after, 0) as u64;
```

Comment:
Just use as u64, the node does not support timestamps before unix epoch anywhere.

Reply:
Ok, nice.